content/blog/using-docker-wrong-my-journey-to-a-better-container.md

```diff
@@ -0,0 +1,53 @@
+
+<https://github.com/johnomernik/maprssss_wrong>
+
+As you can see by the name, I did it wrong.  Now, most folks would just update their code once they realize a mistake; however, I created a second repository because I wanted to share what I did wrong as an example to others.  What did I do wrong?  I did not create a deterministic container (<https://en.wikipedia.org/wiki/Deterministic_system>).  
+
+One of the primary benefits of Docker should be that, if I create a Docker container, I will be able run it on all my systems, and if I hand that container to someone else, it should run the same way, thereby being a deterministic system. Now, let me say there are always caveats, but this should be a large goal for Docker users.
+
+Where did I go wrong? Well, my main issue was using my old code from before the MapR PACC.  In this code, I took a number of directories for StreamSets, including logs, conf (config), and others, and placed it on the local system.  When I ran this on a MapR node in my cluster, it worked great, because all the users that I would use inside the container were also on my MapR nodes running Docker. Thus, I could set up permissions on those local directories, and when the container ran, the user inside the container had proper filesystem permissions to the directories required by StreamSets.
+
+![](/img/docker-host.png)
+
+However, when I shared my repository with a colleague at MapR, they reported it didn’t work on their system. I was baffled: it should work; it’s Docker! That’s the whole point of using containers. Well, my colleague was not running the container on MapR nodes; instead, it was running on a client machine. This client machine did NOT have the same users as the MapR cluster. Thus there were a number of issues that my scripts ran into with creating the configuration directories, settings permissions, and even running StreamSets.  I made assumptions in my original Docker, and by doing so, made a container that ran differently depending on the execution system.
+
+This was an eye-opener for me.  From one perspective, I had tried to get StreamSets running in the PACC by using old knowledge and approaches in an effort to expedite the process. By doing so, I hampered the usability, portability, and shareability of my container. To that end, I’ve created another GitHub repository:
+
```

Comment:
You may want to hotlink the GitHub URL so it reads:
To that end, I've created another [GitHub repository](https://github.com/johnomernik/maprssss).